### PR TITLE
feat(noir): added assert keyword

### LINKED ISSUE #384: Change constrain key word to assert

### PR DESCRIPTION
**Pull request by sirasistant**
# Related issue(s)

Resolves #384

# Description

Adds assert as syntax sugar for constrain. 

## Summary of changes
 - Added assert keyword
 - Added a parser for assertion that generates a ConstrainStatement
 - Added tests

## Test additions / changes

# Checklist

- [x] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [x] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

File: crates/noirc_frontend/src/parser/parser.rs [hunk 1]
```diff
@@ -1214,6 +1223,48 @@ mod test {
         );
     }
 
+    /// The assert statement is syntax sugar for constrain
+    #[test]
```
Comment (kevaundray) [inline crates/noirc_frontend/src/parser/parser.rs:1227]: Nice!

File: crates/noirc_frontend/src/parser/parser.rs [hunk 2]
```diff
@@ -428,12 +428,21 @@ where
...
     choice((
         constrain(expr_parser.clone()),
+        assertion(expr_parser.clone()),
         declaration(expr_parser.clone()),
         assignment(expr_parser.clone()),
         expr_parser.map(Statement::Expression),
     ))
 }
 
+fn assertion<'a, P>(expr_parser: P) -> impl NoirParser<Statement> + 'a
```
Comment (kevaundray) [inline crates/noirc_frontend/src/parser/parser.rs:438]: Another way we could have written it is by doing something like:

```rust
fn assertion<'a, P>(expr_parser: P) -> impl NoirParser<Statement> + 'a
where
    P: ExprParser + 'a,
{
    ignore_then_commit(
        choice((keyword(Keyword::Assert), keyword(Keyword::Constrain))).labelled("statement"),
        expr_parser,
    )
    .map(|expr| Statement::Constrain(ConstrainStatement(expr)))
}
```

Where `choice` essentially means "or".

Then we can remove the `constrain`  function and its tests. We then would not have duplicated code between the assert and constrain test case.

**Review by kevaundray — COMMENTED** [review]
Generally looks good -- left some comments

**Review by kevaundray — APPROVED** [review]

File: crates/noirc_frontend/src/parser/parser.rs [hunk 3]
```diff
@@ -438,7 +438,7 @@ fn constrain<'a, P>(expr_parser: P) -> impl NoirParser<Statement> + 'a
 where
     P: ExprParser + 'a,
 {
-    ignore_then_commit(keyword(Keyword::Constrain).labelled("statement"), expr_parser)
+    ignore_then_commit(choice((keyword(Keyword::Assert), keyword(Keyword::Constrain))).labelled("statement"), expr_parser)
```
Comment (jfecher) [inline crates/noirc_frontend/src/parser/parser.rs:441]: For future note, if there are only two items you can also use `keyword(Keyword::Assert).or(keyword(Keyword::Constrain)`

**Review by jfecher — COMMENTED** [review]
Looks good to me. I'll echo Tom's question about whether we want to go ahead with the `assert(bool)` function call style syntax now as well. We should keep the syntax of the old `constrain` the same either way for backwards compatibility.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**